administration/src/bp-modules/cards/hooks/useCardGenerator.ts

```diff
@@ -25,29 +35,25 @@ const useCardGenerator = (region: Region) => {
   const appToaster = useAppToaster()
 
   const generateCards = useCallback(async () => {
-    try {
-      setState(CardActivationState.loading)
+    let codes: CreateCardsResult[] | undefined
+    setState(CardActivationState.loading)
 
-      const dynamicCodes = cardBlueprints.map(cardBlueprint => {
-        if (projectConfig.activityLogConfig) {
-          new ActivityLog(cardBlueprint).saveToSessionStorage()
-        }
-        return cardBlueprint.generateActivationCode()
-      })
-      const staticCodes = projectConfig.staticQrCodesEnabled
-        ? cardBlueprints.map(cardBlueprints => {
-            return cardBlueprints.generateStaticVerificationCode()
-          })
-        : []
-
-      const pdfDataUri = await generatePdf(dynamicCodes, staticCodes, cardBlueprints, region, projectConfig.pdf)
+    try {
+      const cardInfos = cardBlueprints.map(card => card.generateCardInfo())
+      codes = await createCards(client, projectConfig.projectId, cardInfos, projectConfig.staticQrCodesEnabled)
 
-      const codes = [...dynamicCodes, ...staticCodes]
-      await createCards(client, projectConfig.projectId, codes, region)
+      const pdfDataUri = await generatePdf(codes, cardBlueprints, region, projectConfig.pdf)
 
+      cardBlueprints.forEach(cardBlueprint => new ActivityLog(cardBlueprint).saveToSessionStorage())
       downloadDataUri(pdfDataUri, 'berechtigungskarten.pdf')
       setState(CardActivationState.finished)
     } catch (e) {
+      if (codes !== undefined) {
+        // try rollback
+        try {
+          await deleteCards(client, projectConfig.projectId, region.id, extractCardInfoHashes(codes))
+        } catch {}
```

Comment:
maybe add a sentry report here? We'd then need to set up sentry for web, though.

Reply:
you can create an issue and add a comment here. has not to be done is this task